### DISCUSSION TITLE
Question regarding 'Should you Pin your JavaScript Dependencies?'

### DISCUSSION THREAD

**Discussion by gkranis** [renovatebot/renovate#19402] [Suggest an Idea]
This is not a JS specific question (even though im mentioning package.json for example) and excuse me if I'm missing package manager details. I'm trying to understand the philosophy behind [that article](https://docs.renovatebot.com/dependency-pinning/) before I figure out whats best for us.

The way I see it lockfiles have everything that a "pinned style" package.json have plus pinned indirect dependencies. Their only downside AFAIU is that lockfiles are traditionally generated from scratch each time rather than bumped incrementally package by package. Which leads to many changes grouped together in a single PR...

So not considering the specifics of package managers if there was a way to bump lockfiles incrementally then `[ranged package.json + lockfile]` would be better than the recommended `[pinned package.json + lockfile]`. The former has the advantage that developers can encode legit range limitation in their `package.json` rather than having to instead write them on `renovate.json` (which could be met with resistance....)

So my suspicion/question is that maybe the recommended `[pinned package.json + lockfile]` is recommended not in principle but only because of how upstream pip-compile/npm etc manage lockfiles (or how renovate calls them), so there could be improvement potential....

**Comment by rarkins** [comment]
There's no question in the above three paragraphs. But I can ask one!

I'm not sure what you mean by "developers can encode legit range limitation". For what reason do they need a range, and what makes it "legit"?

**Reply by gkranis** [reply]
A legit range limitation is a developer deciding he doesnt want to migrate to the next major version of a library. In that case he would add a range in package.json . If he had to follow the article's recommendation he would instead have to add that range to renovate.json . In that case he just moved the responsibility to `renovate.json` and nothing changed so this does not look any better than using ranges.


My question is: If there was a package manager (or renovate was smart enough) that allowed you to refresh the lockfile one package at a time would you still recommend not using ranges in requirements.txt ? Because if there was this would cancel the biggest downside of depending on a lockfile which is described in the sentence `Alternatively, maybe someone thinks "This lockfile is probably really out of date and might be missing some essential patches" and decides to yarn upgrade the whole thing in one go`

**Reply by rarkins** [reply]
> A legit range limitation is a developer deciding he doesnt want to migrate to the next major version of a library. In that case he would add a range in package.json . If he had to follow the article's recommendation he would instead have to add that range to renovate.json . In that case he just moved the responsibility to `renovate.json` and nothing changed so this does not look any better than using ranges.

I don't understand.

So your "legit range" is e.g. `^3.0.0` instead of pinning to e.g. `3.5.0`. In both cases you'd get a PR proposing to update to v4 when it comes out. In both cases you could close that PR to ignore it. You haven't moved any responsibility, you've just made it harder to know exactly which 3.x version you're using because you have to look into the lock file to know it's 3.5.0.
 
> My question is: If there was a package manager (or renovate was smart enough) that allowed you to refresh the lockfile one package at a time would you still recommend not using ranges in requirements.txt ? Because if there was this would cancel the biggest downside of depending on a lockfile which is described in the sentence `Alternatively, maybe someone thinks "This lockfile is probably really out of date and might be missing some essential patches" and decides to yarn upgrade the whole thing in one go`

Renovate's rangeStrategy=update-lockfile attempts to do this
